File: LegacySource/settlement/room/infra/importt/ImportThingy.cs
using System;
using System.Collections.Generic;
using snake2d.util.datatypes;

namespace settlement.room.infra.importt
{
    public sealed class ImportThingy
    {
        private readonly ROOM_IMPORT b;
        private readonly ImportTally tally;

        public ImportThingy(ROOM_IMPORT imp, ImportTally tally)
        {
            this.b = imp;
            this.tally = tally;
        }

        public COORDINATE GetReservableSpot(int sx, int sy, RESOURCE res)
        {
            if (tally.capacity[res] - (tally.amount[res]) < 0)
            {
                return null;
            }

            ImportInstance ins = b.Get(sx, sy);
            if (ins == null || ins.Resource != res || Reservable(ins) <= 0)
            {
                ins = null;
                if (b.All().Count == 0)
                    return null;
                int r = RND.rInt(b.All().Count);
                for (int i = 0; i < b.All().Count; i++)
                {
                    ImportInstance ins2 = b.All()[(i + r) % b.All().Count];
                    if (ins2.Resource == res && Reservable(ins2) > 0)
                    {
                        ins = ins2;
                        break;
                    }
                }
            }

            if (ins != null)
            {
                return GetReservableSpot(ins, sx, sy, res);
            }
            return null;
        }

        private COORDINATE GetReservableSpot(ImportInstance i, int sx, int sy, RESOURCE res)
        {
            if (!i.Is(sx, sy))
            {
                sx = i.MX;
                sy = i.MY;
            }
            GUTIL.Filler().Init(this);
            GUTIL.Filler().Filler.Set(sx, sy);
            DIR dir = DIR.ORTHO.rnd();
            int q = 0;
            while (GUTIL.Filler().HasMore())
            {
                COORDINATE c = GUTIL.Filler().Poll();
                if (Reservable(res, c) > 0)
                {
                    GUTIL.Filler().Done();
                    return c;
                }
                q++;
                DIR d = dir;
                for (int k = 0; k < DIR.ORTHO.Size; k++)
                {
                    if (i.Is(c, d))
                        GUTIL.Filler().Fill(c, d);
                    d = d.Next(2);
                }
            }
            GUTIL.Filler().Done();
            GAME.Notify("oh no " + res + " " + i.Resource + " " + q + " " + i.Area + " " + i.MX + " " + i.MY + " " + Reservable(i));
            return null;
        }

        private int Reservable(ImportInstance ins)
        {
            return ins.Capacity - ins.Amount - ins.SpaceReserved;
        }

        private StorageCrate Get(RESOURCE res, COORDINATE c)
        {
            ImportInstance ins = b.Get(c.X, c.Y);
            if (ins == null)
                return null;

            if (ins.Resource != res)
                return null;

            return b.Crate.Get(c.X, c.Y, ins, ins.SData);
        }

        public int Reservable(RESOURCE r, COORDINATE c)
        {
            StorageCrate cr = Get(r, c);
            if (cr == null)
                return 0;
            return cr.StorageReservable;
        }

        public void Reserve(RESOURCE r, COORDINATE c, int amount)
        {
            StorageCrate cr = Get(r, c);
            if (cr == null)
                throw new Exception();
            cr.StorageReserve(amount);
        }

        public int Reserved(RESOURCE r, COORDINATE c)
        {
            StorageCrate cr = Get(r, c);
            if (cr == null)
                return 0;
            return cr.StorageReserved;
        }

        public void Finish(RESOURCE r, COORDINATE c, int amount, TRADE_TYPE type)
        {
            StorageCrate cr = Get(r, c);
            if (cr == null)
                throw new Exception();
            cr.StorageDeposit(amount);
            FACTIONS.Player().Res.Inc(r, type.RType, amount);
        }
    }
}

File: settlement/room/infra/importt/ImportThingy.java
package settlement.room.infra.importt;

import game.GAME;
import game.faction.FACTIONS;
import init.resources.RESOURCE;
import init.trade.TRADE_TYPE;
import settlement.room.main.job.StorageCrate;
import snake2d.util.datatypes.COORDINATE;
import snake2d.util.datatypes.DIR;
import snake2d.util.rnd.RND;
import util.GUTIL;

public final class ImportThingy {

	private final ROOM_IMPORT b;
	private final ImportTally tally;
	
	
	ImportThingy(ROOM_IMPORT imp, ImportTally tally){
		this.b = imp;
		this.tally = tally;
	}
	
	public COORDINATE getReservableSpot(int sx, int sy, RESOURCE res) {

		if (tally.capacity.get(res)-(tally.amount.get(res)) < 0) {
			return null;
		}
			
		
		ImportInstance ins = b.get(sx, sy);
		if (ins == null || ins.resource() != res || reservable(ins) <= 0) {
			ins = null;
			if (b.all().size() == 0)
				return null;
			int r = RND.rInt(b.all().size());
			for (int i = 0; i < b.all().size(); i++) {
				ImportInstance ins2 = b.all().get((i+r)%b.all().size());
				if (ins2.resource() == res && reservable(ins2) > 0) {
					ins = ins2;
					break;
				}
			}
		}
		
		if (ins != null) {
			return getReservableSpot(ins, sx, sy, res);
		}
		return null;
	}
	
	private COORDINATE getReservableSpot(ImportInstance i, int sx, int sy, RESOURCE res) {
		if (!i.is(sx, sy)) {
			sx = i.mX();
			sy = i.mY();
		}
		GUTIL.filler().init(this);
		GUTIL.filler().filler.set(sx, sy);
		DIR dir = DIR.ORTHO.rnd();
		int q = 0;
		while(GUTIL.filler().hasMore()) {
			COORDINATE c = GUTIL.filler().poll();
			if (reservable(res, c) > 0) {
				GUTIL.filler().done();
				return c;
			}
			q++;
			DIR d = dir;
			for (int k = 0; k < DIR.ORTHO.size(); k++) {
				if (i.is(c, d))
					GUTIL.filler().fill(c, d);
				d = d.next(2);
			}
			
		}
		GUTIL.filler().done();
		GAME.Notify("oh no " + res + " " + i.resource() + " " + q + " " + i.area() + " " + i.mX() + " " + i.mY() + " " + reservable(i));
		return null;
	}
	
	private int reservable(ImportInstance ins) {
		return ins.capacity()-ins.amount()-ins.spaceReserved();
	}
	
	private StorageCrate get(RESOURCE res, COORDINATE c) {
		ImportInstance ins = b.get(c.x(), c.y());
		if (ins == null)
			return null;
		
		if (ins.resource() != res)
			return null;
		
		return b.crate.get(c.x(), c.y(), ins, ins.sdata);
	}
	
	public int reservable(RESOURCE r, COORDINATE c) {
		StorageCrate cr = get(r, c);
		if (cr == null)
			return 0;
		return cr.storageReservable();
		
	}

	public void reserve(RESOURCE r, COORDINATE c, int amount) {
		StorageCrate cr = get(r, c);
		if (cr == null)
			throw new RuntimeException();
		cr.storageReserve(amount);;
	}
	
	public int reserved(RESOURCE r, COORDINATE c) {
		StorageCrate cr = get(r, c);
		if (cr == null)
			return 0;
		return cr.storageReserved();
	}
	
	public void finish(RESOURCE r, COORDINATE c, int amount, TRADE_TYPE type) {
		StorageCrate cr = get(r, c);
		if (cr == null)
			throw new RuntimeException();
		cr.storageDeposit(amount);
		FACTIONS.player().res().inc(r, type.rtype, amount);
	}
	
	

	
}
